contracts/vaults/LiquidityLock.sol

```diff
@@ -0,0 +1,161 @@
+pragma solidity ^0.8.0;
+
+import "../../interfaces/IToken.sol";
+import "../../interfaces/IBZx.sol";
+import "../proxies/0_8/Upgradeable_0_8.sol";
+import "@openzeppelin-4.7.0/token/ERC20/utils/SafeERC20.sol";
+
+contract LiquidityLock is Ownable {
+    using SafeERC20 for IERC20;
+
+    event Locked(address indexed user, address indexed iToken, bytes32 claimID, uint256 iTokenAmount, uint256 endTimestamp);
+    event Unlocked(address indexed user, bytes32 claimID);
+    event ClaimableBonus(address indexed token, uint256 amount);
+    event LockCap(address indexed token, uint256 amount);
+    event UpdatedSettings(uint256 discount, uint256 discountAddition, uint256 maxLock);
+    event LockUpdated(bytes32 indexed claimID, uint256 iTokenAmount);
+    IBZx public immutable PROTOCOL;
+    address public immutable TREASURY;
+    mapping(bytes32=>LockedLiquidity) public claims;
+    mapping(address=>uint256) public claimableBonus;
+    mapping(address=>Settings) public lockSettings;
+    mapping(address=>uint256) public totalLocked;
+    mapping(address=>uint256) public lockCap;
+    struct LockedLiquidity {
+        address token;
+        uint256 startTimestamp;
+        uint256 endTimestamp;
+        uint256 iTokenAmount;
+        uint256 originalDeposit;
+        uint256 endAmountReceived;
+    }
+    struct Settings {
+        uint256 minimumDiscount;
+        uint256 discountAdditionPerSecond;
+        uint256 maxLockTime;   
+    }
+    constructor(IBZx p, address t) {
+        PROTOCOL = p;
+        TREASURY = t;
+    }
+
+    function lock(address base, uint256 depositAmount, uint256 endTimestamp) external returns (bytes32 claimID) {
+        Settings memory tokenSettings = lockSettings[base];
+        IToken iToken = IToken(PROTOCOL.underlyingToLoanPool(base));
+        claimID = keccak256(abi.encode(address(iToken),msg.sender,endTimestamp));
+        require(claims[claimID].token == address(0), "already used id");
+        require(block.timestamp <= endTimestamp && endTimestamp-block.timestamp<tokenSettings.maxLockTime, "invalid ending time");
```

Comment:
`endTimestamp-block.timestamp<tokenSettings.maxLockTime` is sufficient enough to check.  it will fail anyway if `block.timestamp <= endTimestamp` is false